### DISCUSSION TITLE
Using postUpgradeTasks to run pod-install to update ios/Podfile.lock in bare React Native  project

### DISCUSSION THREAD

**Discussion by jzaefferer** [renovatebot/renovate#25645] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

37.27.1

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

Very similar to https://github.com/renovatebot/renovate/discussions/10335 I want to run pod install after updating dependencies, to update ios/Podfile.lock.

### What you tried so far.

With the input from https://github.com/renovatebot/renovate/discussions/25612, I've now got this script in my package.json:
```
"renovate:pod-install": "RUBY_VERSION=2.7.8 install-tool ruby && COCOAPODS_VERSION=1.14.2 install-gem cocoapods && yarn --ignore-scripts && cd ios && pod install",
```
I'm hoping to remove the ruby and cocoapods install, since that should already be available. Is it just not in the global PATH and I need to invoke a specific path?

I also created this `renovate-config.js`:
```
module.exports = {
  // this can only be used in a 'global' config: https://docs.renovatebot.com/getting-started/running/#global-config
  allowedPostUpgradeCommands: ['yarn renovate:pod-install'],
  baseBranches: ['renovate-updates'],
};
```
I'm using the `baseBranches` while debugging.

In my renovate.json itself very little changed:
```
"postUpgradeTasks": {
    "commands": ["yarn renovate:pod-install"],
    "fileFilters": ["ios/Podfile.lock"],
    "executionMode": "branch"
  },
```

Now the script runs, but eventually fails, because no node_modules are actually installed. Eventually it outputs this:
```
Error: Cannot find module 'expo/package.json'
Require stack:
- /builds/tonies/react-native/renovate/repos/gitlab/tonies/react-native/ios/[eval]
    at Module._resolveFilename (node:internal/modules/cjs/loader:1048:15)
    at Function.resolve (node:internal/modules/helpers:136:19)
    at [eval]:1:9
    at Script.runInThisContext (node:vm:122:12)
    at Object.runInThisContext (node:vm:296:38)
    at node:internal/process/execution:83:21
    at [eval]-wrapper:6:24
    at runScript (node:internal/process/execution:82:62)
    at evalScript (node:internal/process/execution:104:10)
    at node:internal/main/eval_string:50:3 {
    code: 'MODULE_NOT_FOUND',
  requireStack: [
      '/builds/tonies/react-native/renovate/repos/gitlab/tonies/react-native/ios/[eval]'
  ]
}
[!] Invalid `Podfile` file: cannot load such file -- ./scripts/autolinking.

 #  from /builds/tonies/react-native/renovate/repos/gitlab/tonies/react-native/ios/Podfile:1
 #  -------------------------------------------
 >  require File.join(File.dirname(`node --print \"require.resolve('expo/package.json')\"`), \"scripts/autolinking\")
```
(I stitched this together from stdout and stderror from the debug ndjson file; which I'm trying to parse with a script, but somehow `node` can't decode it correctly)

I seems like I'm pretty close, but since I don't really understand how `renovate` executes `postUpgradeTasks`, I'm stumbling around in the dark. Waiting ~4 minutes for any debugging attempt to dump some output in a GitLab artifact that I then have to download and parse manually is not very effective 😅 

Any ideas on what I'm missing here?

### Relevant debug logs

<details><summary>Log extract with install script mentioned above</summary>

```
Error: Cannot find module 'expo/package.json'
Require stack:
- /builds/tonies/react-native/renovate/repos/gitlab/tonies/react-native/ios/[eval]
    at Module._resolveFilename (node:internal/modules/cjs/loader:1048:15)
    at Function.resolve (node:internal/modules/helpers:136:19)
    at [eval]:1:9
    at Script.runInThisContext (node:vm:122:12)
    at Object.runInThisContext (node:vm:296:38)
    at node:internal/process/execution:83:21
    at [eval]-wrapper:6:24
    at runScript (node:internal/process/execution:82:62)
    at evalScript (node:internal/process/execution:104:10)
    at node:internal/main/eval_string:50:3 {
    code: 'MODULE_NOT_FOUND',
  requireStack: [
      '/builds/tonies/react-native/renovate/repos/gitlab/tonies/react-native/ios/[eval]'
  ]
}
[!] Invalid `Podfile` file: cannot load such file -- ./scripts/autolinking.

 #  from /builds/tonies/react-native/renovate/repos/gitlab/tonies/react-native/ios/Podfile:1
 #  -------------------------------------------
 >  require File.join(File.dirname(`node --print \"require.resolve('expo/package.json')\"`), \"scripts/autolinking\")
```

</details>

full debug log: https://gist.github.com/jzaefferer/ef5a33f8d94c5b1c7e24b0d9bc9435b3

**Comment by rarkins** [comment]
The tools won't necessarily already be available unless you're using the `-full` images. The default images do not come with tools preinstalled.

Also, you should not worry about the "inefficiency" of running `install-tool` because if the same tool/version were already installed then it won't take up any time at all and returns immediately

**Reply by jzaefferer** [reply]
Thanks for those details. That at least confirms my approach is going in the right direction.

Any ideas why `expo/package.json` can't be found? It looks like node_modules is empty, yet `yarn install` says there's nothing to install.

**Reply by viceice** [reply, marked as answer]
you probably use yarn V1 and we install a patched version. you need to disable skip install to get node modules

**Reply by jzaefferer** [reply]
Thanks. Adding `skipInstalls: false,` to my config.js helped with that.

Now some other part of `pod install` is failing 🙄 This feels like I'm going down the wrong rabbit hole.

**Reply by jzaefferer** [reply]
PS: https://docs.renovatebot.com/self-hosted-configuration/#skipinstalls says the default is `null`, which seems misleading if it's implicitly `true`.

Though I don't think I ever would've found this option without your answer here 🤷

**Reply by jzaefferer** [reply]
We're probably going to solve this by pinning the few dependencies we have in `Podfile` and deleting `Podfile.lock` from our git repo.

Thanks @rarkins and @viceice for your help and such a great tool!
